services/housetables/src/main/java/com/linkedin/openhouse/housetables/services/UserTablesServiceImpl.java

```diff
@@ -157,16 +166,69 @@ public void renameUserTable(
   }
 
   @Override
-  public void deleteUserTable(String databaseId, String tableId) {
-    if (!htsJdbcRepository.existsById(
-        UserTableRowPrimaryKey.builder().databaseId(databaseId).tableId(tableId).build())) {
-      throw new NoSuchUserTableException(databaseId, tableId);
+  @Transactional
+  public void deleteUserTable(String databaseId, String tableId, boolean isSoftDeleted) {
+    UserTableRow existingTable =
+        htsJdbcRepository
+            .findById(
+                UserTableRowPrimaryKey.builder().databaseId(databaseId).tableId(tableId).build())
+            .orElseThrow(() -> new NoSuchUserTableException(databaseId, tableId));
+    if (isSoftDeleted) {
+      softDeletedHtsJdbcRepository.save(
+          softDeletedUserTablesMapper.toSoftDeletedUserTableRow(existingTable));
     }
-
     htsJdbcRepository.deleteById(
         UserTableRowPrimaryKey.builder().databaseId(databaseId).tableId(tableId).build());
   }
 
+  @Override
+  @Transactional
+  public UserTableDto restoreUserTable(String databaseId, String tableId, Long deletedAt) {
+    SoftDeletedUserTableRowPrimaryKey softDeletedTableKey =
+        SoftDeletedUserTableRowPrimaryKey.builder()
+            .databaseId(databaseId)
+            .tableId(tableId)
+            .deletedAtMs(deletedAt)
+            .build();
+    SoftDeletedUserTableRow existingSoftDeletedTable =
+        softDeletedHtsJdbcRepository
+            .findById(softDeletedTableKey)
+            .orElseThrow(() -> new NoSuchUserTableException(databaseId, tableId));
+    try {
+      softDeletedHtsJdbcRepository.deleteById(softDeletedTableKey);
```

Comment:
Looks like restoring latest required additional work to find the latest id and then delete.

Reply:
Yes, restoring would be 2 api calls (1 get, 1 restore call). But it'd be bound to 2 calls, neither will be heavyweight because we can sort the get results in the query and . I think that is fine given that restore won't be called often. Let me know if you believe otherwise.